crates/bevy_math/src/shape_sampling.rs

```diff
@@ -0,0 +1,313 @@
+    /// //  and the other is somewhere between -1 and 1.
+    /// println!("{:?}", square.sample_boundary(&mut rand::thread_rng()));
+    /// ```
+    fn sample_boundary<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::Output;
+}
+
+impl ShapeSample for Circle {
+    type Output = Vec2;
+
+    fn sample_interior<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        // https://mathworld.wolfram.com/DiskPointPicking.html
+        let theta = rng.gen_range(0.0..TAU);
+        let r_squared = rng.gen_range(0.0..=(self.radius * self.radius));
+        let r = r_squared.sqrt();
+        Vec2::new(r * theta.cos(), r * theta.sin())
+    }
+
+    fn sample_boundary<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        let theta = rng.gen_range(0.0..TAU);
+        Vec2::new(self.radius * theta.cos(), self.radius * theta.sin())
+    }
+}
+
+impl ShapeSample for Sphere {
+    type Output = Vec3;
+
+    fn sample_interior<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        // https://mathworld.wolfram.com/SpherePointPicking.html
+        let theta = rng.gen_range(0.0..TAU);
+        let phi = rng.gen_range(-1.0_f32..1.0).acos();
+        let r_cubed = rng.gen_range(0.0..=(self.radius * self.radius * self.radius));
+        let r = r_cubed.cbrt();
+        Vec3 {
+            x: r * phi.sin() * theta.cos(),
+            y: r * phi.sin() * theta.sin(),
+            z: r * phi.cos(),
+        }
+    }
+
+    fn sample_boundary<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        let theta = rng.gen_range(0.0..TAU);
+        let phi = rng.gen_range(-1.0_f32..1.0).acos();
+        Vec3 {
+            x: self.radius * phi.sin() * theta.cos(),
+            y: self.radius * phi.sin() * theta.sin(),
+            z: self.radius * phi.cos(),
+        }
+    }
+}
+
+impl ShapeSample for Rectangle {
+    type Output = Vec2;
+
+    fn sample_interior<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        let x = rng.gen_range(-self.half_size.x..=self.half_size.x);
+        let y = rng.gen_range(-self.half_size.y..=self.half_size.y);
+        Vec2::new(x, y)
+    }
+
+    fn sample_boundary<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec2 {
+        let primary_side = rng.gen_range(-1.0..1.0);
+        let other_side = if rng.gen() { -1.0 } else { 1.0 };
+
+        if rng.gen_bool((self.half_size.x / (self.half_size.x + self.half_size.y)) as f64) {
+            Vec2::new(primary_side, other_side) * self.half_size
+        } else {
+            Vec2::new(other_side, primary_side) * self.half_size
+        }
+    }
+}
+
+impl ShapeSample for Cuboid {
+    type Output = Vec3;
+
+    fn sample_interior<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
+        let x = rng.gen_range(-self.half_size.x..=self.half_size.x);
+        let y = rng.gen_range(-self.half_size.y..=self.half_size.y);
+        let z = rng.gen_range(-self.half_size.z..=self.half_size.z);
+        Vec3::new(x, y, z)
+    }
+
+    fn sample_boundary<R: Rng + ?Sized>(&self, rng: &mut R) -> Vec3 {
```

Comment:
As pointed out by IQuick 143 on discord this implementation is currently very wrong